File: server/geocode.py
"""Event → (lat, lon) for the globe.

Two paths: (1) explicit coords from a structured ingestor (USGS) pass straight
through; (2) otherwise we resolve a country mentioned in the subject / affected
entities to a representative centroid. This is a SEED country table (facts —
approximate centroids for major countries); extend toward the full
worldmonitor country-bbox set as needed. Returns (None, None) when nothing matches
— the globe simply doesn't plot that event.
"""

from __future__ import annotations

import re

# country name -> (lat, lon) approximate centroid. Seed set; extend freely.
COUNTRY_CENTROID: dict[str, tuple[float, float]] = {
    "united states": (39.8, -98.6),
    "usa": (39.8, -98.6),
    "u.s.": (39.8, -98.6),
    "china": (35.9, 104.2),
    "russia": (61.5, 105.3),
    "ukraine": (48.4, 31.2),
    "israel": (31.0, 34.9),
    "gaza": (31.4, 34.4),
    "palestine": (31.9, 35.2),
    "iran": (32.4, 53.7),
    "iraq": (33.2, 43.7),
    "saudi arabia": (23.9, 45.1),
    "united kingdom": (54.0, -2.0),
    "uk": (54.0, -2.0),
    "britain": (54.0, -2.0),
    "france": (46.6, 2.2),
    "germany": (51.2, 10.4),
    "spain": (40.5, -3.7),
    "italy": (41.9, 12.6),
    "japan": (36.2, 138.3),
    "south korea": (36.5, 127.8),
    "north korea": (40.3, 127.5),
    "india": (22.0, 79.0),
    "pakistan": (30.4, 69.3),
    "taiwan": (23.7, 121.0),
    "turkey": (39.0, 35.2),
    "egypt": (26.8, 30.8),
    "brazil": (-14.2, -51.9),
    "mexico": (23.6, -102.5),
    "canada": (56.1, -106.3),
    "australia": (-25.3, 133.8),
    "south africa": (-30.6, 22.9),
    "nigeria": (9.1, 8.7),
    "venezuela": (6.4, -66.6),
    "yemen": (15.6, 48.0),
    "syria": (34.8, 38.9),
    "lebanon": (33.9, 35.9),
    "afghanistan": (33.9, 67.7),
    "qatar": (25.3, 51.2),
    "united arab emirates": (23.4, 53.8),
    "uae": (23.4, 53.8),
    "poland": (51.9, 19.1),
    "netherlands": (52.1, 5.3),
    "switzerland": (46.8, 8.2),
    "sweden": (60.1, 18.6),
    "indonesia": (-0.8, 113.9),
    "philippines": (12.9, 121.8),
    "vietnam": (14.1, 108.3),
    "thailand": (15.9, 100.99),
    "greece": (39.1, 21.8),
    "hong kong": (22.3, 114.2),
    "singapore": (1.35, 103.8),
    "argentina": (-38.4, -63.6),
    "libya": (26.3, 17.2),
}

# match longest names first so "south korea" wins over "korea"
_SORTED = sorted(COUNTRY_CENTROID, key=len, reverse=True)


def geocode_text(text: str) -> tuple[float | None, float | None]:
    """Find the first country mentioned (word-boundary) and return its centroid."""
    low = (text or "").lower()
    for name in _SORTED:
        if re.search(rf"(?<![a-z]){re.escape(name)}(?![a-z])", low):
            return COUNTRY_CENTROID[name]
    return None, None


def geocode_event(
    subject: str,
    affected_entities: list | None = None,
    explicit_latlon: tuple[float | None, float | None] | None = None,
) -> tuple[float | None, float | None]:
    """Resolve coordinates: explicit ingestor coords win; else scan subject then
    entity names for a country."""
    if (
        explicit_latlon
        and explicit_latlon[0] is not None
        and explicit_latlon[1] is not None
    ):
        return explicit_latlon

    lat, lon = geocode_text(subject)
    if lat is not None:
        return lat, lon

    for ent in affected_entities or []:
        name = ent.get("name") if isinstance(ent, dict) else str(ent)
        lat, lon = geocode_text(name or "")
        if lat is not None:
            return lat, lon
    return None, None

File: finscrape/absorbed/geo/newsglobe/lib/geocode.ts
import { LOCATIONS } from "./countries";
import { cityMapping } from "city-timezones";

// --- Build combined location lookup (city-timezones + manual overrides) ---

const CITY_COORDS = new Map<string, { lat: number; lng: number }>();

// First load all 7000+ cities from city-timezones
for (const city of cityMapping) {
  const key = city.city.toLowerCase();
  // Prefer larger cities (higher population) when duplicates exist
  const existing = CITY_COORDS.get(key);
  if (!existing || (city.pop && city.pop > 0)) {
    CITY_COORDS.set(key, { lat: city.lat, lng: city.lng });
  }
  // Also index by "city, country" for disambiguation
  const fullKey = `${city.city}, ${city.country}`.toLowerCase();
  CITY_COORDS.set(fullKey, { lat: city.lat, lng: city.lng });
}

// Then overlay manual LOCATIONS (countries, regions, custom entries take priority)
for (const [key, coords] of Object.entries(LOCATIONS)) {
  CITY_COORDS.set(key, coords);
}

// --- Nominatim geocoding with persistent cache ---

const nominatimCache = new Map<string, { lat: number; lng: number } | null>();
let lastNominatimCall = 0;
const NOMINATIM_MIN_INTERVAL = 1100;

async function nominatimGeocode(
  query: string
): Promise<{ lat: number; lng: number } | null> {
  const cacheKey = query.toLowerCase().trim();
  if (nominatimCache.has(cacheKey)) return nominatimCache.get(cacheKey)!;

  // Rate limit
  const now = Date.now();
  const wait = NOMINATIM_MIN_INTERVAL - (now - lastNominatimCall);
  if (wait > 0) await new Promise((r) => setTimeout(r, wait));
  lastNominatimCall = Date.now();

  try {
    const url = `https://nominatim.openstreetmap.org/search?q=${encodeURIComponent(query)}&format=json&limit=1`;
    const res = await fetch(url, {
      headers: { "User-Agent": "NewsGlobe/1.0 (news visualization app)" },
    });
    if (!res.ok) {
      nominatimCache.set(cacheKey, null);
      return null;
    }
    const data = (await res.json()) as { lat: string; lon: string }[];
    if (data.length === 0) {
      nominatimCache.set(cacheKey, null);
      return null;
    }
    const result = {
      lat: parseFloat(data[0].lat),
      lng: parseFloat(data[0].lon),
    };
    nominatimCache.set(cacheKey, result);
    return result;
  } catch {
    nominatimCache.set(cacheKey, null);
    return null;
  }
}

// --- Location keyword extraction ---

const REGIONS = new Set([
  "middle east", "europe", "european", "africa", "african", "asia", "asian",
  "latin america", "eu", "nato",
]);

const DEMONYMS = new Set([
  "american", "canadian", "british", "french", "german", "italian", "spanish",
  "dutch", "swiss", "irish", "greek", "turkish", "ukrainian", "russian",
  "chinese", "japanese", "korean", "indian", "iranian", "syrian", "saudi",
  "israeli", "palestinian", "egyptian", "brazilian", "australian",
]);

const COUNTRIES = new Set([
  "united states", "usa", "u.s.", "america", "canada", "mexico", "brazil",
  "argentina", "colombia", "chile", "peru", "venezuela", "cuba",
  "united kingdom", "uk", "britain", "england", "scotland", "france",
  "germany", "italy", "spain", "portugal", "netherlands", "belgium",
  "switzerland", "austria", "poland", "czech", "sweden", "norway", "denmark",
  "finland", "ireland", "greece", "turkey", "romania", "hungary", "ukraine",
  "russia", "china", "japan", "south korea", "north korea", "india",
  "pakistan", "bangladesh", "indonesia", "thailand", "vietnam", "philippines",
  "malaysia", "singapore", "taiwan", "myanmar", "cambodia", "sri lanka",
  "nepal", "afghanistan", "iraq", "iran", "syria", "saudi arabia", "israel",
  "palestine", "gaza", "lebanon", "jordan", "yemen", "uae", "dubai", "qatar",
  "kuwait", "egypt", "libya", "tunisia", "algeria", "morocco", "nigeria",
  "south africa", "kenya", "ethiopia", "ghana", "tanzania", "uganda", "sudan",
  "congo", "somalia", "senegal", "mozambique", "zimbabwe", "australia",
  "new zealand",
]);

const FALSE_POSITIVES = new Set(["jordan"]);

interface Match {
  key: string;
  specificity: number;
  position: number;
}

function getSpecificity(key: string): number {
  if (REGIONS.has(key)) return 1;
  if (DEMONYMS.has(key)) return 1.5;
  if (COUNTRIES.has(key)) return 2;
  return 3; // cities, states, specific places
}

const locationKeys = Array.from(CITY_COORDS.keys());

function extractCandidates(text: string): Match[] {
  const lower = text.toLowerCase();
  const matches: Match[] = [];

  for (const key of locationKeys) {
    const idx = lower.indexOf(key);
    if (idx === -1) continue;

    const before = lower[idx - 1];
    const after = lower[idx + key.length];
    const wb = (ch: string | undefined) =>
      !ch || /[\s,.\-—:;!?'"()[\]{}\/]/.test(ch);

    if (!wb(before) || !wb(after)) continue;

    if (FALSE_POSITIVES.has(key)) {
      const ctx = lower.slice(Math.max(0, idx - 5), idx).trim();
      if (!/\b(in|of|to|from)\b/.test(ctx)) continue;
    }

    matches.push({ key, specificity: getSpecificity(key), position: idx });
  }

  matches.sort((a, b) => {
    if (b.specificity !== a.specificity) return b.specificity - a.specificity;
    return a.position - b.position;
  });

  return matches;
}

// --- Fast geocode (static table only, no Nominatim, no async) ---

export function geocodeFast(
  text: string
): { lat: number; lng: number } | null {
  const candidates = extractCandidates(text);
  if (candidates.length === 0) return null;
  return staticFallback(candidates[0].key);
}

// --- Main geocode (async, Nominatim-enhanced) ---

export async function geocode(
  text: string
): Promise<{ lat: number; lng: number } | null> {
  const candidates = extractCandidates(text);
  if (candidates.length === 0) return null;

  const best = candidates[0];

  // If best match is already a specific city/state (specificity 3) and is
  // in our static map, use Nominatim to refine it for better precision
  // For country-level matches (specificity 2), also use Nominatim
  // For regions/demonyms (< 2), just use static map (Nominatim won't help)
  if (best.specificity >= 2) {
    // Check if already cached in Nominatim (instant)
    const cached = nominatimCache.get(best.key.toLowerCase().trim());
    if (cached !== undefined) {
      if (cached) {
        return jitter(cached, best.key);
      }
      return staticFallback(best.key);
    }

    // Not cached — try Nominatim (will take ~1.1s due to rate limit)
    const result = await nominatimGeocode(best.key);
    if (result) {
      return jitter(result, best.key);
    }
  }

  return staticFallback(best.key);
}

function staticFallback(
  key: string
): { lat: number; lng: number } | null {
  const loc = CITY_COORDS.get(key);
  if (!loc) return null;
  return jitter(loc, key);
}

// Coastal cities where jitter should push inland only
const COASTAL_BIAS: Record<string, { latDir: number; lngDir: number }> = {
  mumbai: { latDir: 1, lngDir: 1 },     // push NE (inland)
  chennai: { latDir: 1, lngDir: 1 },
  "hong kong": { latDir: 1, lngDir: 1 },
  shanghai: { latDir: 1, lngDir: -1 },
  tokyo: { latDir: 1, lngDir: -1 },
  sydney: { latDir: -1, lngDir: -1 },
  "new york": { latDir: 1, lngDir: 1 },
  miami: { latDir: 1, lngDir: 1 },
  "los angeles": { latDir: 1, lngDir: 1 },
  "san francisco": { latDir: 1, lngDir: 1 },
  london: { latDir: 0, lngDir: 0 },      // inland enough
  dubai: { latDir: 1, lngDir: 1 },
  "tel aviv": { latDir: 1, lngDir: 1 },
  singapore: { latDir: 1, lngDir: 0 },
  istanbul: { latDir: 0, lngDir: 1 },
  lagos: { latDir: 1, lngDir: 1 },
  jakarta: { latDir: -1, lngDir: 1 },
  manila: { latDir: 1, lngDir: 1 },
  lima: { latDir: -1, lngDir: 1 },
  "rio de janeiro": { latDir: -1, lngDir: 1 },
  "cape town": { latDir: -1, lngDir: 1 },
  lisbon: { latDir: 1, lngDir: 1 },
  barcelona: { latDir: 1, lngDir: -1 },
  "gaza": { latDir: 1, lngDir: 1 },
};

function jitter(
  coords: { lat: number; lng: number },
  locationKey?: string
): { lat: number; lng: number } {
  const amount = 0.08; // ~9km, reduced from 0.15
  const bias = locationKey ? COASTAL_BIAS[locationKey.toLowerCase()] : undefined;

  let latOff: number;
  let lngOff: number;

  if (bias) {
    // Bias jitter toward inland direction
    latOff = bias.latDir === 0
      ? (Math.random() - 0.5) * amount
      : Math.abs(Math.random() * amount) * bias.latDir;
    lngOff = bias.lngDir === 0
      ? (Math.random() - 0.5) * amount
      : Math.abs(Math.random() * amount) * bias.lngDir;
  } else {
    latOff = (Math.random() - 0.5) * amount;
    lngOff = (Math.random() - 0.5) * amount;
  }

  return {
    lat: coords.lat + latOff,
    lng: coords.lng + lngOff,
  };
}
